src/System.Runtime.Analyzers/Core/DoNotRaiseReservedExceptionTypes.cs

```diff
@@ -35,14 +70,74 @@ public abstract class DoNotRaiseReservedExceptionTypesAnalyzer : DiagnosticAnaly
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(TooGenericRule, ReservedRule); 
+
+        public abstract TLanguageKindEnum ObjectCreationExpressionKind { get; }
+
+        public abstract SyntaxNode GetTypeSyntaxNode(TObjectCreationExpressionSyntax node);
 
         public override void Initialize(AnalysisContext analysisContext)
-        { 
-            
+        {
+                
+            analysisContext.RegisterCompilationStartAction(
+            compilationStartContext =>
+            {
+                HashSet<INamedTypeSymbol> tooGenericExceptionSymbols = CreateSymbolSet(compilationStartContext.Compilation, s_tooGenericExceptions);
+                HashSet<INamedTypeSymbol> reservedExceptionSymbols = CreateSymbolSet(compilationStartContext.Compilation, s_reservedExceptions); ;             
+
+                if (tooGenericExceptionSymbols.Count == 0 && reservedExceptionSymbols.Count == 0)
+                {
+                    return;
+                }
+
+                compilationStartContext.RegisterSyntaxNodeAction(
+                syntaxNodeContext =>
+                {
+                    Analyze(syntaxNodeContext, tooGenericExceptionSymbols, reservedExceptionSymbols);
+                },
+                ObjectCreationExpressionKind);
+            });    
+        }
+
+        private static HashSet<INamedTypeSymbol> CreateSymbolSet(Compilation compilation, IEnumerable<string> exceptionNames)
+        {
+            HashSet< INamedTypeSymbol> set = null;
+            foreach (var exp in exceptionNames)
+            {
+                var symbol = compilation.GetTypeByMetadataName(exp);
+                if (symbol == null)
+                {
+                    continue;
+                }
+                if (set == null)
+                {
+                    set = new HashSet<INamedTypeSymbol>();
+                }
+                set.Add(symbol);
+            }
+            return set;
+        }
+
+        private void Analyze(SyntaxNodeAnalysisContext context, HashSet<INamedTypeSymbol> tooGenericExceptionSymbols, HashSet<INamedTypeSymbol> reservedExceptionSymbols)
+        {
+            var objectCreationNode = (TObjectCreationExpressionSyntax) context.Node;
+            var targetType = GetTypeSyntaxNode(objectCreationNode);
+
+            // TODO: can this be null in error condition?
+            Debug.Assert(targetType != null);
+
+            var  typeSymbol = (INamedTypeSymbol)context.SemanticModel.GetSymbolInfo(targetType).Symbol;
+            if (tooGenericExceptionSymbols.Contains(typeSymbol))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(TooGenericRule, targetType.GetLocation(), "CALLER", typeSymbol.ToDisplayString(new SymbolDisplayFormat(typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces))));
```

Comment:
Yes. I will need to modify the message so it fit the Roslyn style (it is copied from FxCop and really long and verbose). Will fix with that change in next commit.